grpc/shared_conn.go

```diff
@@ -18,68 +24,221 @@ type OnTrackCB func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver)
+		// locks.
+		sc.logger = moduleLogger.Sublogger("conn")
+	}
+
+	if sc.resOnTrackCBs == nil {
+		// Same initilization argument as above with the logger.
+		sc.resOnTrackCBs = make(map[resource.Name]OnTrackCB)
+	}
+
+	sc.peerConnMu.Lock()
+	defer sc.peerConnMu.Unlock()
+
+	if sc.peerConn != nil {
+		sc.logger.Warn("SharedConn is being reset with an active peer connection.")
+		utils.UncheckedError(sc.peerConn.Close())
+		sc.peerConn = nil
+	}
+
+	peerConn, err := webrtc.NewPeerConnection(webrtc.Configuration{})
+	if err != nil {
+		sc.logger.Warnw("Unable to create optional peer connection for module. Ignoring.", "err", err)
+		return
+	}
+
+	sc.peerConn = peerConn
+	sc.peerConnReady, err = rpc.ConfigureForRenegotiation(peerConn, sc.logger.AsZap())
+	if err != nil {
+		sc.logger.Warnw("Unable to create optional renegotiation channel for module. Ignoring.", "err", err)
+		utils.UncheckedError(sc.peerConn.Close())
+		sc.peerConn = nil
+		return
+	}
+
+	sc.peerConn.OnTrack(func(trackRemote *webrtc.TrackRemote, rtpReceiver *webrtc.RTPReceiver) {
+		name, err := resource.NewFromString(trackRemote.StreamID())
+		if err != nil {
+			sc.logger.Errorw("StreamID did not parse as a ResourceName", "sharedConn", fmt.Sprintf("%p", sc), "streamID", trackRemote.StreamID())
+			return
+		}
+		sc.resOnTrackMu.Lock()
+		onTrackCB, ok := sc.resOnTrackCBs[name]
+		sc.resOnTrackMu.Unlock()
+		if !ok {
+			sc.logger.Errorw("Callback not found for StreamID", "sharedConn", fmt.Sprintf("%p", sc), "streamID", trackRemote.StreamID())
+			return
+		}
+		onTrackCB(trackRemote, rtpReceiver)
+	})
+}
+
+// GenerateEncodedOffer creates a WebRTC offer that's JSON + base64 encoded. If an error is
+// returned, `SharedConn.PeerConn` will return nil until a following `Reset`.
+func (sc *SharedConn) GenerateEncodedOffer() (string, error) {
+	success := false
+	defer func() {
+		// Only keep the `peerConn` pointer active if we believe the connection is useable.
+		if !success {
+			sc.peerConn = nil
+		}
+	}()
+
+	pc := sc.PeerConn()
+	if pc == nil {
+		return "", errors.New("peer connections disabled")
+	}
+
+	offer, err := pc.CreateOffer(nil)
+	if err != nil {
+		return "", err
+	}
+
+	if err = pc.SetLocalDescription(offer); err != nil {
+		return "", err
+	}
+
+	<-webrtc.GatheringCompletePromise(pc)
+	success = true
+	return rpc.EncodeSDP(pc.LocalDescription())
+}
+
+// ProcessEncodedAnswer sets the remote description to the answer and waits for the connection to be
+// ready as per the negotiation channel being opened. The answer is expected to be JSON + base64
+// encoded. If an error is returned, `SharedConn.PeerConn` will return nil until a following
+// `Reset`.
+func (sc *SharedConn) ProcessEncodedAnswer(encodedAnswer string) error {
+	success := false
+	defer func() {
+		// Only keep the `peerConn` pointer active if we believe the connection is useable.
+		if !success {
+			sc.peerConn = nil
```

Comment:
If we are serializing access to `sc.peerConn`, do we also need to take that lock durring this nil assignment?